File: ethcore/src/trace/trace.rs
// Copyright 2015, 2016 Ethcore (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

//! Tracing datatypes.
use util::{U256, Bytes, Address, FixedHash};
use util::rlp::*;
use util::sha3::Hashable;
use action_params::ActionParams;
use basic_types::LogBloom;

/// `Call` result.
#[derive(Debug, Clone, PartialEq, Default)]
pub struct CallResult {
	/// Gas used by call.
	pub gas_used: U256,
	/// Call Output.
	pub output: Bytes,
}

impl Encodable for CallResult {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_list(2);
		s.append(&self.gas_used);
		s.append(&self.output);
	}
}

impl Decodable for CallResult {
	fn decode<D>(decoder: &D) -> Result<Self, DecoderError> where D: Decoder {
		let d = decoder.as_rlp();
		let res = CallResult {
			gas_used: try!(d.val_at(0)),
			output: try!(d.val_at(1)),
		};

		Ok(res)
	}
}

/// `Create` result.
#[derive(Debug, Clone, PartialEq)]
pub struct CreateResult {
	/// Gas used by create.
	pub gas_used: U256,
	/// Code of the newly created contract.
	pub code: Bytes,
	/// Address of the newly created contract.
	pub address: Address,
}

impl Encodable for CreateResult {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_list(3);
		s.append(&self.gas_used);
		s.append(&self.code);
		s.append(&self.address);
	}
}

impl Decodable for CreateResult {
	fn decode<D>(decoder: &D) -> Result<Self, DecoderError> where D: Decoder {
		let d = decoder.as_rlp();
		let res = CreateResult {
			gas_used: try!(d.val_at(0)),
			code: try!(d.val_at(1)),
			address: try!(d.val_at(2)),
		};

		Ok(res)
	}
}

/// Description of a _call_ action, either a `CALL` operation or a message transction.
#[derive(Debug, Clone, PartialEq)]
pub struct Call {
	/// The sending account.
	pub from: Address,
	/// The destination account.
	pub to: Address,
	/// The value transferred to the destination account.
	pub value: U256,
	/// The gas available for executing the call.
	pub gas: U256,
	/// The input data provided to the call.
	pub input: Bytes,
}

impl From<ActionParams> for Call {
	fn from(p: ActionParams) -> Self {
		Call {
			from: p.sender,
			to: p.address,
			value: p.value.value(),
			gas: p.gas,
			input: p.data.unwrap_or_else(Vec::new),
		}
	}
}

impl Encodable for Call {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_list(5);
		s.append(&self.from);
		s.append(&self.to);
		s.append(&self.value);
		s.append(&self.gas);
		s.append(&self.input);
	}
}

impl Decodable for Call {
	fn decode<D>(decoder: &D) -> Result<Self, DecoderError> where D: Decoder {
		let d = decoder.as_rlp();
		let res = Call {
			from: try!(d.val_at(0)),
			to: try!(d.val_at(1)),
			value: try!(d.val_at(2)),
			gas: try!(d.val_at(3)),
			input: try!(d.val_at(4)),
		};

		Ok(res)
	}
}

impl Call {
	/// Returns call action bloom.
	/// The bloom contains from and to addresses.
	pub fn bloom(&self) -> LogBloom {
		LogBloom::from_bloomed(&self.from.sha3())
			.with_bloomed(&self.to.sha3())
	}
}

/// Description of a _create_ action, either a `CREATE` operation or a create transction.
#[derive(Debug, Clone, PartialEq)]
pub struct Create {
	/// The address of the creator.
	pub from: Address,
	/// The value with which the new account is endowed.
	pub value: U256,
	/// The gas available for the creation init code.
	pub gas: U256,
	/// The init code.
	pub init: Bytes,
}

impl From<ActionParams> for Create {
	fn from(p: ActionParams) -> Self {
		Create {
			from: p.sender,
			value: p.value.value(),
			gas: p.gas,
			init: p.code.unwrap_or_else(Vec::new),
		}
	}
}

impl Encodable for Create {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_list(4);
		s.append(&self.from);
		s.append(&self.value);
		s.append(&self.gas);
		s.append(&self.init);
	}
}

impl Decodable for Create {
	fn decode<D>(decoder: &D) -> Result<Self, DecoderError> where D: Decoder {
		let d = decoder.as_rlp();
		let res = Create {
			from: try!(d.val_at(0)),
			value: try!(d.val_at(1)),
			gas: try!(d.val_at(2)),
			init: try!(d.val_at(3)),
		};

		Ok(res)
	}
}

impl Create {
	/// Returns bloom create action bloom.
	/// The bloom contains only from address.
	pub fn bloom(&self) -> LogBloom {
		LogBloom::from_bloomed(&self.from.sha3())
	}
}

/// Description of an action that we trace; will be either a call or a create.
#[derive(Debug, Clone, PartialEq)]
pub enum Action {
	/// It's a call action.
	Call(Call),
	/// It's a create action.
	Create(Create),
}

impl Encodable for Action {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_list(2);
		match *self {
			Action::Call(ref call) => {
				s.append(&0u8);
				s.append(call);
			},
			Action::Create(ref create) => {
				s.append(&1u8);
				s.append(create);
			}
		}
	}
}

impl Decodable for Action {
	fn decode<D>(decoder: &D) -> Result<Self, DecoderError> where D: Decoder {
		let d = decoder.as_rlp();
		let action_type: u8 = try!(d.val_at(0));
		match action_type {
			0 => d.val_at(1).map(Action::Call),
			1 => d.val_at(1).map(Action::Create),
			_ => Err(DecoderError::Custom("Invalid action type.")),
		}
	}
}

impl Action {
	/// Returns action bloom.
	pub fn bloom(&self) -> LogBloom {
		match *self {
			Action::Call(ref call) => call.bloom(),
			Action::Create(ref create) => create.bloom(),
		}
	}
}

/// The result of the performed action.
#[derive(Debug, Clone, PartialEq)]
pub enum Res {
	/// Successful call action result.
	Call(CallResult),
	/// Successful create action result.
	Create(CreateResult),
	/// Failed call.
	FailedCall,
	/// Failed create.
	FailedCreate,
}

impl Encodable for Res {
	fn rlp_append(&self, s: &mut RlpStream) {
		match *self {
			Res::Call(ref call) => {
				s.begin_list(2);
				s.append(&0u8);
				s.append(call);
			},
			Res::Create(ref create) => {
				s.begin_list(2);
				s.append(&1u8);
				s.append(create);
			},
			Res::FailedCall => {
				s.begin_list(1);
				s.append(&2u8);
			},
			Res::FailedCreate => {
				s.begin_list(1);
				s.append(&3u8);
			}
		}
	}
}

impl Decodable for Res {
	fn decode<D>(decoder: &D) -> Result<Self, DecoderError> where D: Decoder {
		let d = decoder.as_rlp();
		let action_type: u8 = try!(d.val_at(0));
		match action_type {
			0 => d.val_at(1).map(Res::Call),
			1 => d.val_at(1).map(Res::Create),
			2 => Ok(Res::FailedCall),
			3 => Ok(Res::FailedCreate),
			_ => Err(DecoderError::Custom("Invalid result type.")),
		}
	}
}

#[derive(Debug, Clone, PartialEq)]
/// A trace; includes a description of the action being traced and sub traces of each interior action.
pub struct Trace {
	/// The number of EVM execution environments active when this action happened; 0 if it's
	/// the outer action of the transaction.
	pub depth: usize,
	/// The action being performed.
	pub action: Action,
	/// The sub traces for each interior action performed as part of this call.
	pub subs: Vec<Trace>,
	/// The result of the performed action.
	pub result: Res,
}

impl Encodable for Trace {
	fn rlp_append(&self, s: &mut RlpStream) {
		s.begin_list(4);
		s.append(&self.depth);
		s.append(&self.action);
		s.append(&self.subs);
		s.append(&self.result);
	}
}

impl Decodable for Trace {
	fn decode<D>(decoder: &D) -> Result<Self, DecoderError> where D: Decoder {
		let d = decoder.as_rlp();
		let res = Trace {
			depth: try!(d.val_at(0)),
			action: try!(d.val_at(1)),
			subs: try!(d.val_at(2)),
			result: try!(d.val_at(3)),
		};

		Ok(res)
	}
}

impl Trace {
	/// Returns trace bloom.
	pub fn bloom(&self) -> LogBloom {
		self.subs.iter().fold(self.action.bloom(), |b, s| b | s.bloom())
	}
}

#[cfg(test)]
mod tests {
	use util::{Address, U256, FixedHash};
	use util::rlp::{encode, decode};
	use util::sha3::Hashable;
	use trace::trace::{Call, CallResult, Create, Res, Action, Trace};

	#[test]
	fn traces_rlp() {
		let trace = Trace {
			depth: 2,
			action: Action::Call(Call {
				from: Address::from(1),
				to: Address::from(2),
				value: U256::from(3),
				gas: U256::from(4),
				input: vec![0x5]
			}),
			subs: vec![
				Trace {
					depth: 3,
					action: Action::Create(Create {
						from: Address::from(6),
						value: U256::from(7),
						gas: U256::from(8),
						init: vec![0x9]
					}),
					subs: vec![],
					result: Res::FailedCreate
				}
			],
			result: Res::Call(CallResult {
				gas_used: U256::from(10),
				output: vec![0x11, 0x12]
			})
		};

		let encoded = encode(&trace);
		let decoded: Trace = decode(&encoded);
		assert_eq!(trace, decoded);
	}

	#[test]
	fn traces_bloom() {
		let trace = Trace {
			depth: 2,
			action: Action::Call(Call {
				from: Address::from(1),
				to: Address::from(2),
				value: U256::from(3),
				gas: U256::from(4),
				input: vec![0x5]
			}),
			subs: vec![
				Trace {
					depth: 3,
					action: Action::Create(Create {
						from: Address::from(6),
						value: U256::from(7),
						gas: U256::from(8),
						init: vec![0x9]
					}),
					subs: vec![],
					result: Res::FailedCreate
				}
			],
			result: Res::Call(CallResult {
				gas_used: U256::from(10),
				output: vec![0x11, 0x12]
			})
		};

		let bloom = trace.bloom();

		// right now only addresses are bloomed
		assert!(bloom.contains_bloomed(&Address::from(1).sha3()));
		assert!(bloom.contains_bloomed(&Address::from(2).sha3()));
		assert!(!bloom.contains_bloomed(&Address::from(20).sha3()));
		assert!(bloom.contains_bloomed(&Address::from(6).sha3()));
	}
}

File: js-old/src/jsonrpc/interfaces/trace.js
// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import { Address, BlockNumber, Data, Hash, CallRequest } from '../types';
import { withPreamble, Dummy, fromDecimal } from '../helpers';

const SECTION_FILTERING = 'Transaction-Trace Filtering';
const SECTION_ADHOC = 'Ad-hoc Tracing';

export default withPreamble(`

The trace module is for getting a deeper insight into transaction processing.
It includes two sets of calls; the transaction trace filtering API and the ad-hoc tracing API.

**Note:** In order to use these API Parity must be fully synced with flags \`$ parity --tracing on\`.

## The Ad-hoc Tracing API

The ad-hoc tracing API allows you to perform a number of different diagnostics on calls or transactions,
either historical ones from the chain or hypothetical ones not yet mined. The diagnostics include:

- \`trace\` **Transaction trace**. An equivalent trace to that in the previous section.
- \`vmTrace\` **Virtual Machine execution trace**. Provides a full trace of the VM's state throughout the execution of the transaction, including for any subcalls.
- \`stateDiff\` **State difference**. Provides information detailing all altered portions of the Ethereum state made due to the execution of the transaction.

There are three means of providing a transaction to execute; either providing the same information as when making
a call using \`eth_call\` (see \`trace_call\`), through providing raw, signed, transaction data as when using
\`eth_sendRawTransaction\` (see \`trace_rawTransaction\`) or simply a transaction hash for a previously mined
transaction (see \`trace_replayTransaction\`). In the latter case, your node must be in archive mode or the
transaction should be within the most recent 1000 blocks.

## The Transaction-Trace Filtering API

These APIs allow you to get a full *externality* trace on any transaction executed throughout the Parity chain.
Unlike the log filtering API, you are able to search and filter based only upon address information.
Information returned includes the execution of all \`CREATE\`s, \`SUICIDE\`s and all variants of \`CALL\` together
with input data, output data, gas usage, amount transferred and the success status of each individual action.

### \`traceAddress\` field

The \`traceAddress\` field of all returned traces, gives the exact location in the call trace [index in root,
index in first \`CALL\`, index in second \`CALL\`, ...].

i.e. if the trace is:
\`\`\`
A
  CALLs B
    CALLs G
  CALLs C
    CALLs G
\`\`\`
then it should look something like:

\`[ {A: []}, {B: [0]}, {G: [0, 0]}, {C: [1]}, {G: [1, 0]} ]\`

`, {
  block: {
    section: SECTION_FILTERING,
    desc: 'Returns traces created at given block.',
    params: [
      {
        type: BlockNumber,
        desc: 'Integer of a block number, or the string `\'earliest\'`, `\'latest\'` or `\'pending\'`.',
        example: fromDecimal(3068185)
      }
    ],
    returns: {
      type: Array,
      desc: 'Block traces.',
      example: [
        {
          action: {
            callType: 'call',
            from: '0xaa7b131dc60b80d3cf5e59b5a21a666aa039c951',
            gas: '0x0',
            input: '0x',
            to: '0xd40aba8166a212d6892125f079c33e6f5ca19814',
            value: '0x4768d7effc3fbe'
          },
          blockHash: '0x7eb25504e4c202cf3d62fd585d3e238f592c780cca82dacb2ed3cb5b38883add',
          blockNumber: 3068185,
          result: {
            gasUsed: '0x0',
            output: '0x'
          },
          subtraces: 0,
          traceAddress: [],
          transactionHash: '0x07da28d752aba3b9dd7060005e554719c6205c8a3aea358599fc9b245c52f1f6',
          transactionPosition: 0,
          type: 'call'
        },
        new Dummy('...')
      ]
    }
  },

  filter: {
    section: SECTION_FILTERING,
    desc: 'Returns traces matching given filter',
    params: [
      {
        type: Object,
        desc: 'The filter object',
        details: {
          fromBlock: {
            type: BlockNumber,
            desc: 'From this block.',
            optional: true
          },
          toBlock: {
            type: BlockNumber,
            desc: 'To this block.',
            optional: true
          },
          fromAddress: {
            type: Array,
            desc: 'Sent from these addresses.',
            optional: true
          },
          toAddress: {
            type: Address,
            desc: 'Sent to these addresses.',
            optional: true
          }
        },
        example: {
          fromBlock: fromDecimal(3068100),
          toBlock: fromDecimal(3068200),
          toAddress: ['0x8bbB73BCB5d553B5A556358d27625323Fd781D37']
        }
      }
    ],
    returns: {
      type: Array,
      desc: 'Traces matching given filter',
      example: [
        {
          action: {
            callType: 'call',
            from: '0x32be343b94f860124dc4fee278fdcbd38c102d88',
            gas: '0x4c40d',
            input: '0x',
            to: '0x8bbb73bcb5d553b5a556358d27625323fd781d37',
            value: '0x3f0650ec47fd240000'
          },
          blockHash: '0x86df301bcdd8248d982dbf039f09faf792684e1aeee99d5b58b77d620008b80f',
          blockNumber: 3068183,
          result: {
            gasUsed: '0x0',
            output: '0x'
          },
          subtraces: 0,
          traceAddress: [],
          transactionHash: '0x3321a7708b1083130bd78da0d62ead9f6683033231617c9d268e2c7e3fa6c104',
          transactionPosition: 3,
          type: 'call'
        },
        new Dummy('...')
      ]
    }
  },

  get: {
    section: SECTION_FILTERING,
    desc: 'Returns trace at given position.',
    params: [
      {
        type: Hash,
        desc: 'Transaction hash.',
        example: '0x17104ac9d3312d8c136b7f44d4b8b47852618065ebfa534bd2d3b5ef218ca1f3'
      },
      {
        type: Array,
        desc: 'Index positions of the traces.',
        example: ['0x0']
      }
    ],
    returns: {
      type: Object,
      desc: 'Trace object',
      example: {
        action: {
          callType: 'call',
          from: '0x1c39ba39e4735cb65978d4db400ddd70a72dc750',
          gas: '0x13e99',
          input: '0x16c72721',
          to: '0x2bd2326c993dfaef84f696526064ff22eba5b362',
          value: '0x0'
        },
        blockHash: '0x7eb25504e4c202cf3d62fd585d3e238f592c780cca82dacb2ed3cb5b38883add',
        blockNumber: 3068185,
        result: {
          gasUsed: '0x183',
          output: '0x0000000000000000000000000000000000000000000000000000000000000001'
        },
        subtraces: 0,
        traceAddress: [0],
        transactionHash: '0x17104ac9d3312d8c136b7f44d4b8b47852618065ebfa534bd2d3b5ef218ca1f3',
        transactionPosition: 2,
        type: 'call'
      }
    }
  },

  transaction: {
    section: SECTION_FILTERING,
    desc: 'Returns all traces of given transaction',
    params: [
      {
        type: Hash,
        desc: 'Transaction hash',
        example: '0x17104ac9d3312d8c136b7f44d4b8b47852618065ebfa534bd2d3b5ef218ca1f3'
      }
    ],
    returns: {
      type: Array,
      desc: 'Traces of given transaction',
      example: [
        {
          action: {
            callType: 'call',
            from: '0x1c39ba39e4735cb65978d4db400ddd70a72dc750',
            gas: '0x13e99',
            input: '0x16c72721',
            to: '0x2bd2326c993dfaef84f696526064ff22eba5b362',
            value: '0x0'
          },
          blockHash: '0x7eb25504e4c202cf3d62fd585d3e238f592c780cca82dacb2ed3cb5b38883add',
          blockNumber: 3068185,
          result: {
            gasUsed: '0x183',
            output: '0x0000000000000000000000000000000000000000000000000000000000000001'
          },
          subtraces: 0,
          traceAddress: [0],
          transactionHash: '0x17104ac9d3312d8c136b7f44d4b8b47852618065ebfa534bd2d3b5ef218ca1f3',
          transactionPosition: 2,
          type: 'call'
        },
        new Dummy('...')
      ]
    }
  },

  call: {
    section: SECTION_ADHOC,
    desc: 'Executes the given call and returns a number of possible traces for it.',
    params: [
      {
        type: CallRequest,
        desc: 'Call options, same as `eth_call`.',
        example: new Dummy('{ ... }')
      },
      {
        type: Array,
        desc: 'Type of trace, one or more of: `"vmTrace"`, `"trace"`, `"stateDiff"`.',
        example: ['trace']
      },
      {
        type: BlockNumber,
        optional: true,
        desc: 'Integer of a block number, or the string `\'earliest\'`, `\'latest\'` or `\'pending\'`.'
      }
    ],
    returns: {
      type: Array,
      desc: 'Block traces',
      example: {
        output: '0x',
        stateDiff: null,
        trace: [
          {
            action: new Dummy('{ ... }'),
            result: {
              gasUsed: '0x0',
              output: '0x'
            },
            subtraces: 0,
            traceAddress: [],
            type: 'call'
          }
        ],
        vmTrace: null
      }
    }
  },

  rawTransaction: {
    section: SECTION_ADHOC,
    desc: 'Traces a call to `eth_sendRawTransaction` without making the call, returning the traces',
    params: [
      {
        type: Data,
        desc: 'Raw transaction data.',
        example: '0xd46e8dd67c5d32be8d46e8dd67c5d32be8058bb8eb970870f072445675058bb8eb970870f072445675'
      },
      {
        type: Array,
        desc: 'Type of trace, one or more of: `"vmTrace"`, `"trace"`, `"stateDiff"`.',
        example: ['trace']
      }
    ],
    returns: {
      type: Object,
      desc: 'Block traces.',
      example: {
        output: '0x',
        stateDiff: null,
        trace: [
          {
            action: new Dummy('{ ... }'),
            result: {
              gasUsed: '0x0',
              output: '0x'
            },
            subtraces: 0,
            traceAddress: [],
            type: 'call'
          }
        ],
        vmTrace: null
      }
    }
  },

  replayTransaction: {
    section: SECTION_ADHOC,
    desc: 'Replays a transaction, returning the traces.',
    params: [
      {
        type: Hash,
        desc: 'Transaction hash.',
        example: '0x02d4a872e096445e80d05276ee756cefef7f3b376bcec14246469c0cd97dad8f'
      },
      {
        type: Array,
        desc: 'Type of trace, one or more of: `"vmTrace"`, `"trace"`, `"stateDiff"`.',
        example: ['trace']
      }
    ],
    returns: {
      type: Object,
      desc: 'Block traces.',
      example: {
        output: '0x',
        stateDiff: null,
        trace: [
          {
            action: new Dummy('{ ... }'),
            result: {
              gasUsed: '0x0',
              output: '0x'
            },
            subtraces: 0,
            traceAddress: [],
            type: 'call'
          }
        ],
        vmTrace: null
      }
    }
  }
});
